pkg/stream/consumer.go

```diff
@@ -148,11 +148,9 @@ func (c *Consumer) process(ctx context.Context, msg *Message) bool {
 		c.handleError(ctx, err, "an error occurred during the consume operation")
 	}
 
-	if ack {
-		return true
+	if !ack {
+		c.retry(ctx, msg)
 	}
 
-	c.retry(ctx, msg)
-
-	return ack
+	return ack || isHeadOfLineBlocking(c.input)
```

Comment:
I'm wondering if the consumer should be aware of these "implementation" details. Meaning that the consumer has to know about how the input is handling acknowledgement. If the next input has again a different acknowledgement pattern, it will get quite ugly here.

Maybe we should let the input decide what to do with a positive or negative ack. If we would change the interface to:
```golang
type AcknowledgeableInput interface {
	Input
	Ack(ctx context.Context, msg *Message, ack bool) error
	AckBatch(ctx context.Context, msgs []*Message, acks []bool) error
}
```
The SQS input can send the acknowledgement based on the ack value and the kafka input can just always send the commit. So the consumer would call Ack every time on the input not just in the success case. I guess this would make it easier, leaner and cleaner..

Reply:
makes sense, pushed :)